File: 1D Dynamic Programming/5. 最长回文子串/5. 最长回文子串.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        res = ""
        maxlen = 0

        for i in range(len(s)):
            # 偶數，注意此處不是判斷s的長度，而是判斷"回文串的長度是奇或偶數"
            # if len(s) % 2 == 0:
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                temp_len = r - l + 1
                if temp_len > maxlen:
                    res = s[l : r + 1]
                    maxlen = temp_len
                l -= 1
                r += 1
            # 奇數
            # else:
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                temp_len = r - l + 1
                if temp_len > maxlen:
                    res = s[l : r + 1]
                    maxlen = temp_len
                l -= 1
                r += 1      

        return res

File: 1D Dynamic Programming/5. 最长回文子串/5. 最长回文子串.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        string res = "";
        int maxlen = 0;

        for(int i = 0;i < s.size();i++){
            // 回文串為奇數(擇一發生)
            int l = i, r = i;
            while(l >= 0 && r < s.size() && s[l] == s[r]){
                if(r - l + 1 > maxlen){
                    maxlen = r - l + 1;
                    res = s.substr(l, maxlen);
                }
                l --;
                r ++;
            }

            // 回文串為偶數(擇一發生)
            l = i, r = i + 1;
            while(l >= 0 && r < s.size() && s[l] == s[r]){
                if(r - l + 1 > maxlen){
                    maxlen = r - l + 1;
                    res = s.substr(l, maxlen);
                }
                l --;
                r ++;
            }

        }